// erine/core/classes/Base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErineClient = void 0;
const discord_js_1 = require("discord.js");
const Core_1 = require("./Core");
const Skyfold_1 = require("./Skyfold");
class ErineClient extends discord_js_1.Client {
    coreback;
    skyfold;
    _options;
    constructor(options) {
        super(options);
        this.skyfold = new Skyfold_1.Skyfold({ client: this });
        this.coreback = new Core_1.Coreback(this);
        this._options = options;
        this.init();
    }
    async load_commands(dir) {
        this.skyfold.load_commands(dir);
    }
    async load_events(dir) {
        this.skyfold.load_events(dir);
    }
    async load_interactions(dir) {
        this.skyfold.load_interactions(dir);
    }
    init() {
        let m = require('../events/messageCreate').event;
        let i = require('../events/interactionCreate').event;
        this.on(m.name, (message) => m.code(message, this));
        this.on(i.name, (int) => i.code(int, this));
    }
}
exports.ErineClient = ErineClient;
