// erine/core/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = exports.Plugins = exports.Errors = exports.ParamsBuilder = exports.HybridBuilder = exports.HybridGroup = exports.PrefixBuilder = exports.ErineClient = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./classes/Base");
Object.defineProperty(exports, "ErineClient", { enumerable: true, get: function () { return Base_1.ErineClient; } });
const PrefixBuilder_1 = require("./classes/builders/PrefixBuilder");
Object.defineProperty(exports, "PrefixBuilder", { enumerable: true, get: function () { return PrefixBuilder_1.PrefixBuilder; } });
const HybridBuilder_1 = require("./classes/builders/HybridBuilder");
Object.defineProperty(exports, "HybridBuilder", { enumerable: true, get: function () { return HybridBuilder_1.HybridBuilder; } });
const HybridGroupBuilder_1 = require("./classes/builders/HybridGroupBuilder");
Object.defineProperty(exports, "HybridGroup", { enumerable: true, get: function () { return HybridGroupBuilder_1.HybridGroup; } });
const ParamsBuilder_1 = require("./classes/builders/ParamsBuilder");
Object.defineProperty(exports, "ParamsBuilder", { enumerable: true, get: function () { return ParamsBuilder_1.ParamsBuilder; } });
const Context_1 = require("./classes/Context");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return Context_1.Context; } });
const Core_1 = require("./classes/Core");
Object.defineProperty(exports, "Plugins", { enumerable: true, get: function () { return Core_1.Plugins; } });
const Errors = tslib_1.__importStar(require("./classes/Errors"));
exports.Errors = Errors;
