// erine/core/classes/builders/ParamsBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParamsBuilder = void 0;
const discord_js_1 = require("discord.js");
class ParamsBuilder {
    strings;
    numbers;
    params;
    constructor() {
        this.strings = [];
        this.numbers = [];
        this.params = [];
    }
    addString(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.String;
        this.params.push(param);
        return this;
    }
    addNumber(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.Number;
        this.params.push(param);
        return this;
    }
    addBoolean(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.Boolean;
        this.params.push(param);
        return this;
    }
    addMember(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.User;
        this.params.push(param);
        return this;
    }
    addChannel(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.Channel;
        this.params.push(param);
        return this;
    }
    addRole(param) {
        param.type = discord_js_1.ApplicationCommandOptionType.Role;
        this.params.push(param);
        return this;
    }
}
exports.ParamsBuilder = ParamsBuilder;
